currency/currency_storage.go

```diff
@@ -0,0 +1,721 @@
+	s.fxRates.mtx.Lock()
+	defer s.fxRates.mtx.Unlock()
+	rates, err := s.fiatExchangeMarkets.GetCurrencyData(
+		s.baseCurrency.String(),
+		s.fiatCurrencies.Strings())
+	if err != nil {
+		return err
+	}
+	return s.updateExchangeRates(rates)
+}
+
+// UpdateForeignExchangeRates sets exchange rates on the FX map
+func (s *Storage) updateExchangeRates(m map[string]float64) error {
+	err := s.fxRates.Update(m)
+	if err != nil {
+		return err
+	}
+
+	if s.path != "" {
+		return s.WriteCurrencyDataToFile(s.path, false)
+	}
+	return nil
+}
+
+// SetupCryptoProvider sets congiguration parameters and starts a new instance
+// of the currency analyser
+func (s *Storage) SetupCryptoProvider(settings coinmarketcap.Settings) error {
+	if settings.APIkey == "" ||
+		settings.APIkey == "key" ||
+		settings.AccountPlan == "" ||
+		settings.AccountPlan == "accountPlan" {
+		return errors.New("currencyprovider error api key or plan not set in config.json")
+	}
+
+	s.currencyAnalysis = new(coinmarketcap.Coinmarketcap)
+	s.currencyAnalysis.SetDefaults()
+	s.currencyAnalysis.Setup(settings)
+
+	return nil
+}
+
+// GetTotalMarketCryptocurrencies returns the total seeded market
+// cryptocurrencies
+func (s *Storage) GetTotalMarketCryptocurrencies() (Currencies, error) {
+	if !s.currencyCodes.HasData() {
+		return nil, errors.New("market currency codes not populated")
+	}
+	return s.currencyCodes.GetCurrencies(), nil
+}
+
+// IsDefaultCurrency returns if a currency is a default currency
+func (s *Storage) IsDefaultCurrency(c Code) bool {
+	t, _ := GetTranslation(c)
+	for _, d := range s.defaultFiatCurrencies {
+		if d.Match(c) || d.Match(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// IsDefaultCryptocurrency returns if a cryptocurrency is a default
+// cryptocurrency
+func (s *Storage) IsDefaultCryptocurrency(c Code) bool {
+	t, _ := GetTranslation(c)
+	for _, d := range s.defaultCryptoCurrencies {
+		if d.Match(c) || d.Match(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// IsFiatCurrency returns if a currency is part of the enabled fiat currency
+// list
+func (s *Storage) IsFiatCurrency(c Code) bool {
+	if c.Item.Role != Unset {
+		return c.Item.Role == Fiat
+	}
+
+	t, _ := GetTranslation(c)
+	for _, d := range s.fiatCurrencies {
+		if d.Match(c) || d.Match(t) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// IsCryptocurrency returns if a cryptocurrency is part of the enabled
+// cryptocurrency list
+func (s *Storage) IsCryptocurrency(c Code) bool {
+	if c.Item.Role != Unset {
+		return c.Item.Role == Cryptocurrency
+	}
+
+	t, _ := GetTranslation(c)
+	for _, d := range s.cryptocurrencies {
+		if d.Match(c) || d.Match(t) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// NewCode validates string against currency list and returns a currency
+// code
+func (s *Storage) NewCode(newCode string) Code {
+	return s.currencyCodes.Register(newCode)
+}
+
+// NewValidFiatCode inserts a new code and updates the fiat currency list
```

Comment:
What is the meaning of Valid here? I think it is meant to say that the provide code (as string) should be treated as "valid"... ie ignore some type of validity tests ... Perhaps AddNewCode or RegisterNewCode() is a more intuitive func name?

Reply:
This assigns a currency role "fiat currency" into the *Item. NewCode is blind and disregards currency roles. When we get currency data from the foreign exchange provider it is *usually* a valid fiat. 

This does need to be commented better to explain the difference though.